src/swarmpal/toolboxes/secs/dsecs_algorithms.py

```diff
@@ -0,0 +1,1895 @@
+    # Map B-field points to Rsecs
+    thetaFP = np.arcsin(
+        np.sqrt(rsecs / rb) * np.sin(thetaB)
+    )  # co-latitude of the footpoints, mapped to the northern hemisphere
+    ind = np.nonzero(thetaB > np.pi / 2)
+    thetaFP[ind] = np.pi - thetaFP[ind]
+    # Horizontal distances as  cosine of co-latitude in the SECS-centered system
+    # See Eq. (A5) and Fig. 14 of Vanhamäki et al.(2003)
+    tmp = np.cos(thetaFP) * np.cos(thetaSECS) + np.sin(thetaFP) * np.sin(
+        thetaSECS
+    ) * np.cos(phiSECS - phiB)
+    minD = rsecs * np.arccos(np.max(tmp))
+
+    # Calculate the theta-angles of the points used in the integration
+    # make sure t is ROW vector
+    t = auto.sub_points_along_fieldline(thetaSECS, rsecs, L, minD)
+
+    # xyz - coordinates of the integration points (= locations of the current elements)
+    # Make sure these are ROW vectors
+    x = L * np.sin(t) ** 3 * np.cos(phiSECS)
+    y = L * np.sin(t) ** 3 * np.sin(phiSECS)
+    z = -L * np.sin(t) ** 2 * np.cos(t)  ###why minus here?
+
+    # xyz  components of the current elements
+    dlx = -np.diff(x)
+    dly = -np.diff(y)
+    dlz = -np.diff(z)
+
+    # xyz-coordinates of the mid-points
+    x = (x[:-1] + x[1:]) / 2
+    y = (y[:-1] + y[1:]) / 2
+    z = (z[:-1] + z[1:]) / 2
+
+    # |distances between current elements and field points|^3
+    diffx = (np.expand_dims(x0, -1) - x).T  # (x0[:,np.newaxis] -x).T
+    diffy = (np.expand_dims(y0, -1) - x).T  # (y0[:,np.newaxis] -y).T
+    diffz = (np.expand_dims(z0, -1) - x).T  # (z0[:,np.newaxis] -z).T
+
+    tt = np.sqrt(diffx * diffx + diffy * diffy + diffz * diffz)
+    root = tt * tt * tt
+
+    # Remove singularity by setting a lower lomit to the distance between current elements and field points.
+    # Use the small number
+    root = np.where(root < smallr * smallr * smallr, smallr, root)
+
+    # xyz components of the magnetic field
+    # now there is a common factor mu0/(4*pi) = 1e.7
+    # If scaling factors as in [A] , radii in [km] and magnetic field in [nT] --> extra factor of 1e6
+    # Make sure the products are (Nb, Nint) matrices and sum along the rows to integrate
+
+    xsum = (dly * diffz.T - dlz * diffy.T) / root.T
+    ysum = (dlx * diffz.T - dlz * diffx.T) / root.T
+    zsum = (dlx * diffy.T - dly * diffx.T) / root.T
+
+    bx = 0.1 * np.sum(xsum.T).T
+    by = 0.1 * np.sum(ysum.T).T
+    bz = 0.1 * np.sum(zsum.T).T
+
+    return bx, by, bz
+
+
+# @jit(parallel=True,nopython=True)
+def _calc_root(x, y, z):
+    """_summary_
+
+    Parameters
+    ----------
+    x : ndarray
+        _description_
+    y : ndarray
+        _description_
+    z : ndarray
+        _description_
+
+    Returns
+    -------
+    ndarray
+        _description_
+    """
+
+    tt = np.sqrt(x * x + y * y + z * z)
+    root = tt * tt * tt
+
+    return root
+
+
+def get_data_slices(
+    t1=dt.datetime(2016, 3, 18, 11, 3, 0),
+    t2=dt.datetime(2016, 3, 18, 11, 40, 0),
```

Comment:
flake8-bugbear complains about this with:
```
B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
```

I don't think this actually causes an issue in this case. We could ignore it, but I think it would be cleaner anyway to simply not provide defaults for `t1`, `t2`, since the defaults only make sense for the test case.

Reply:
I'll remove that. I think it's just leftover from debugging the initial code.